### PR TITLE
[python] Add a task to create a cartesian product of mixture densities

### PR DESCRIPTION
**Pull request by mreboud**

### CODE REVIEW

File: python/eos/tasks.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 # this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 # Place, Suite 330, Boston, MA  02111-1307  USA
 
+from pypmc import density
```
Comment (mreboud) [inline python/eos/tasks.py:18]: Arf, this automatic behavior is really boring. This has nothing to do there. VSCode tries to be too smart...

File: src/scripts/eos-analysis [hunk 2]
```diff
@@ -62,6 +62,7 @@ Lists the named prior PDFs defined within the scope of this analysis file.
     )
     parser_list_priors.set_defaults(cmd = cmd_list_priors)
 
+
```
Comment (mreboud) [inline src/scripts/eos-analysis:65]: I added a couple of new lines for readability.

File: python/eos/data/pypmc.py [hunk 3]
```diff
@@ -155,6 +156,49 @@ def create(path, density):
         with open(os.path.join(path, 'description.yaml'), 'w') as description_file:
             yaml.dump(description, description_file, default_flow_style=False)
 
+    @staticmethod
+    def _cartesian_product(densityA, densityB):
+        """ Compute the cartesian product of two mixture densities. The means and covariances
```
Comment (dvandyk) [inline python/eos/data/pypmc.py:161]: Is this computed or constructed? Nit picking, but if other changes crop up, worthwhile to address?

File: python/eos/data/pypmc.py [hunk 4]
```diff
@@ -155,6 +156,49 @@ def create(path, density):
...
+        of the components are copied and concatenated and the weights are multiplied. Note that
+        this product is not commutative. Only Gaussian mixtures are supported.
```
Comment (dvandyk) [inline python/eos/data/pypmc.py:163]: ``... not commutative in the sense that the order of densities affects the order of parameters in the output density.``

File: python/eos/data/pypmc.py [hunk 5]
```diff
@@ -155,6 +156,49 @@ def create(path, density):
...
+                        raise RuntimeError('Unsupported type of MixtureDensity component: {}'.format(type(cB)))
+            else:
+                raise RuntimeError('Unsupported type of MixtureDensity component: {}'.format(type(cA)))
+        return pypmc.density.mixture.MixtureDensity(components, weights / _np.sum(weights))
+
+    @staticmethod
+    def cartesian_product(densities):
+        """ Compute the cartesian product of a list of mixture densities. The means and covariances
+        of the components are copied and concatenated and the weights are multiplied. Note that
+        this product is not commutative. Only Gaussian mixtures are supported.
```
Comment (dvandyk) [inline python/eos/data/pypmc.py:190]: Ditto (on the commutative)!

File: python/eos/tasks.py [hunk 6]
```diff
@@ -100,10 +100,28 @@ def find_clusters(posterior, base_directory='./', threshold=2.0, K_g=1, analysis
     eos.info(f'Created mixture density with {len(density.components)} components')
     eos.data.MixtureDensity.create(output_path, density)
 
+def mixture_product(posterior, posteriors, base_directory='./', analysis_file=None):
```
Comment (dvandyk) [inline python/eos/tasks.py:103]: I like the name!

File: python/eos/tasks.py [hunk 7]
```diff
@@ -100,10 +100,28 @@ def find_clusters(posterior, base_directory='./', threshold=2.0, K_g=1, analysis
     eos.info(f'Created mixture density with {len(density.components)} components')
     eos.data.MixtureDensity.create(output_path, density)
```
Comment (dvandyk) [inline python/eos/tasks.py:102]: Can you add a bit more space? I think two empty lines give a better visual separation between free standing functions.

File: python/eos/tasks.py [hunk 8]
```diff
@@ -100,10 +100,28 @@ def find_clusters(posterior, base_directory='./', threshold=2.0, K_g=1, analysis
...
+    """
+    Compute the cartesian product of the densities listed in posteriors. Note that this product is not commutative.
+
+    The input densities are read from EOS_BASE_DIRECTORY/POSTERIOR_i/pmc, where POSTERIOR_i is listed in posteriors.
+    The output density will be stored in EOS_BASE_DIRECTORY/POSTERIOR/product.
+
+    :param posterior: The name of the posterior.
+    :type posterior: str
+    :param posteriors: The list of names of the posteriors whose mixture densities will be concatenated.
+    :type posteriors: iterable of str
```
Comment (dvandyk) [inline python/eos/tasks.py:113]: ``posterior`` vs ``posteriors`` is likely to cause trouble, but I have not good (short!) alternative names.

File: python/eos/tasks.py [hunk 9]
```diff
@@ -100,10 +100,28 @@ def find_clusters(posterior, base_directory='./', threshold=2.0, K_g=1, analysis
...
+    :param base_directory: The base directory for the storage of data files. Can also be set via the EOS_BASE_DIRECTORY environment variable.
+    :type base_directory: str, optional
+    """
+
+    densities = [eos.data.PMCSampler(os.path.join(base_directory, p, 'pmc')).density() for p in posteriors]
+    output_path = os.path.join(base_directory, posterior, 'product')
+    eos.data.MixtureDensity.create(output_path, eos.data.MixtureDensity.cartesian_product(densities))
```
Comment (dvandyk) [inline python/eos/tasks.py:121]: Some additional space below?

File: python/eos/tasks.py [hunk 10]
```diff
@@ -126,30 +144,35 @@ def sample_pmc(analysis_file, posterior, base_directory='./', step_N=500, steps=
...
-    _set_log_file(output_path, 'log', mode='a' if continue_sampling else 'w')
+    _set_log_file(output_path, 'log', mode='a' if (initial_proposal == "pmc") else 'w')
     if type(analysis_file) is not eos.AnalysisFile:
         _analysis_file = eos.AnalysisFile(analysis_file)
     else:
         _analysis_file = analysis_file
     analysis = _analysis_file.analysis(posterior)
     rng = _np.random.mtrand.RandomState(1701)
-    if continue_sampling:
+    if (initial_proposal == "clusters"):
```
Comment (dvandyk) [inline python/eos/tasks.py:162]: Consistency: using single quotes ``'`` everywhere except here.

File: python/eos/tasks.py [hunk 11]
```diff
@@ -126,30 +144,35 @@ def sample_pmc(analysis_file, posterior, base_directory='./', step_N=500, steps=
...
+        initial_density = eos.data.MixtureDensity(os.path.join(base_directory, posterior, 'clusters')).density()
+    elif (initial_proposal == "pmc"):
```
Comment (dvandyk) [inline python/eos/tasks.py:164]: Ditto

File: python/eos/tasks.py [hunk 12]
```diff
@@ -126,30 +144,35 @@ def sample_pmc(analysis_file, posterior, base_directory='./', step_N=500, steps=
...
         previous_sampler = eos.data.PMCSampler(os.path.join(base_directory, posterior, 'pmc'))
-        initial_proposal = previous_sampler.density()
+        initial_density = previous_sampler.density()
+    elif (initial_proposal == "product"):
```
Comment (dvandyk) [inline python/eos/tasks.py:167]: Ditto

File: python/eos/tasks.py (hunk 10 above)
Comment (dvandyk) [inline python/eos/tasks.py:162]: Why the parantheses? That's C++ style, not Python.

File: python/eos/tasks.py [hunk 13]
```diff
@@ -126,30 +144,35 @@ def sample_pmc(analysis_file, posterior, base_directory='./', step_N=500, steps=
...
     else:
-        initial_proposal = eos.data.MixtureDensity(os.path.join(base_directory, posterior, 'clusters')).density()
+        eos.error("Could not initialize proposal in sample_pmc: argument {} is not supported.".format(initial_proposal))
 
-    samples, weights, proposal = analysis.sample_pmc(initial_proposal, step_N=step_N, steps=steps, final_N=final_N,
+    samples, weights, proposal = analysis.sample_pmc(initial_density, step_N=step_N, steps=steps, final_N=final_N,
                                                      rng=rng, final_perplexity_threshold=perplexity_threshold, weight_threshold=weight_threshold)
 
-    if continue_sampling:
+    if (initial_proposal == "pmc"):
```
Comment (dvandyk) [inline python/eos/tasks.py:175]: Ditto (parentheses and quotes)

File: src/scripts/eos-analysis (hunk 2 above)
Comment (dvandyk) [reply src/scripts/eos-analysis:65]: Thanks!

**Review by dvandyk — APPROVED** [review]

File: python/eos/tasks.py (hunk 8 above)
Comment (mreboud) [reply python/eos/tasks.py:113]: This should not affect the user since those are mandatory arguments

**Comment by mreboud** [comment]
There is still something weird going on, don't merge :)

**Pull request merged**